willow/plugins/pillow.py

```diff
@@ -246,4 +254,144 @@ def to_buffer_rgba(self):
         return RGBAImageBuffer(image.size, image.tobytes())
 
 
-willow_image_classes = [PillowImage]
+class PillowAnimatedImage(Image):
+    def __init__(self, frames):
+        self.frames = frames
+
+    @classmethod
+    def check(cls):
+        _PIL_Image()
+
+    @classmethod
+    def is_format_supported(cls, image_format):
+        return is_format_supported(image_format)
+
+    @Image.operation
+    def get_size(self):
+        return self.frames[0].get_size()
+
+    @Image.operation
+    def get_frame_count(self):
+        return len(self.frames)
+
+    @Image.operation
+    def has_alpha(self):
+        return self.frames[0].has_alpha()
+
+    @Image.operation
+    def has_animation(self):
+        return self.get_frame_count() > 1
+
+    @Image.operation
+    def resize(self, size):
+        return PillowAnimatedImage([frame.resize(size) for frame in self.frames])
+
+    @Image.operation
+    def crop(self, rect):
+        return PillowAnimatedImage([frame.crop(rect) for frame in self.frames])
+
+    @Image.operation
+    def rotate(self, angle):
+        return PillowAnimatedImage([frame.rotate(angle) for frame in self.frames])
+
+    @Image.operation
+    def set_background_color_rgb(self, color):
+        return PillowAnimatedImage([frame.set_background_color_rgb(color) for frame in self.frames])
+
+    @Image.operation
+    def save_as_jpeg(self, f, quality=85, optimize=False, progressive=False):
+        if self.has_animation():
+            pass  # TODO: Raise warning
+
+        return self.frames[0].save_as_jpeg(f, quality=quality, optimize=optimize, progressive=progressive)
+
+    @Image.operation
+    def save_as_png(self, f, optimize=False):
+        if self.has_animation():
+            pass  # TODO: Raise warning
+
+        return self.frames[0].save_as_png(f, optimize=optimize)
+
+    @Image.operation
+    def save_as_gif(self, f):
+        image = self.frames[0].image
+        frames = self.frames
+
+        # All gif files use either the L or P mode but we sometimes convert them
+        # to RGB/RGBA to improve the quality of resizing. We must make sure that
+        # they are converted back before saving.
+        if image.mode not in ['L', 'P']:
+            frames = [
+                frame.convert('P', palette=_PIL_Image().ADAPTIVE)
+                for frame in frames
+            ]
+
+        params = {
+            'save_all': True,
+            'duration': image.info['duration'],
+            'append_images': [frame.image for frame in frames[1:]]
+        }
+
+        if 'transparency' in image.info:
+            params['transparency'] = image.info['transparency']
```

Comment:
I could not find any reference of variability in frames transparency, but if it turns out to be true, this is another place to consider. also why not user `has_alpha()` as in the image operation?

Reply:
Only information I could find on this is in the [GIF spec](https://www.w3.org/Graphics/GIF/spec-gif89a.txt)

Section 23 says that the fields which define transparency are assigned per graphics block, so it's possible that these settings can change per-frame but even on different areas of a single frame as well.

To check if Pillow supports multiple GIF palettes, I tried opening an image with multiple palettes and saving it again.

Before:
![SmallFullColourGIF](https://user-images.githubusercontent.com/1093808/66197337-18f78880-e692-11e9-8eb9-8ba309b0e784.gif)

After:
![pillow-saved](https://user-images.githubusercontent.com/1093808/66197355-20b72d00-e692-11e9-861d-1d11c191c57f.gif)

I think the answer to that is clear! And looking at the Pillow source code, looks like only one palette gets used in the `_save` function so it's not possible for it to have different palettes for each block: https://github.com/python-pillow/Pillow/blob/master/src/PIL/GifImagePlugin.py#L507

(Image source https://en.wikipedia.org/wiki/GIF#/media/File:SmallFullColourGIF.gif)

Reply:
To demonstrate a tool that can actually do this image blocking/pallette generation automatically, I generated a new GIF using the following website: https://ezgif.com/

![ezgif com-gif-maker](https://user-images.githubusercontent.com/1093808/66200259-8dcdc100-e698-11e9-8ff8-a358ae8ea85c.gif)

This clearly has lower quality than the original, but also it's 6x larger! (315KB up from 54KB)

Trying out other formats, this APNG is 79KB and looks much better:

![ezgif com-apng-maker](https://user-images.githubusercontent.com/1093808/66200443-02a0fb00-e699-11e9-9f27-1974396b6bd3.png)

I can't show them here, but a high quality WEBP comes out at 115KB and a 70% quality WEBP (which looks roughly the same as the GIF above) is 40KB

All modern browsers except IE11 supports at least one of these formats ([apng](https://caniuse.com/#feat=apng ), [webp](https://caniuse.com/#feat=webp)). So I think it would be worth adding support for these into Willow.